pkg/analyzer/nodes/node.go

```diff
@@ -103,22 +104,51 @@ func extractFilesFromDirectory(root string, matcher matcher.PrefixMatcher) (*fil
 	m := metrics.FileExtractionMetrics{}
 	// TODO(ROX-13771): Use `range matcher.GetCommonPrefixDirs()` again after fixing.
 	for _, dir := range []string{"etc/", "usr/share/rpm", "var/lib/rpm", "usr/share/buildinfo"} {
-		if err := n.addFiles(filepath.FromSlash(dir), matcher, &m); err != nil {
-			return nil, errors.Wrapf(err, "failed to match filesMap at %q (at %q)", dir, n.root)
+		if err := n.addFiles(ctx, filepath.FromSlash(dir), matcher, &m); err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				err = errors.Wrapf(err, "operation took longer than expected and was interrupted")
+			}
+			return nil, errors.Wrapf(err, "failed to extract files from %q at root %q", dir, n.root)
 		}
 	}
 	m.Emit()
 	return n, nil
 }
 
+func walkDirWithContext(ctx context.Context, dir string, fn fs.WalkDirFunc) error {
+	// Buffering so we can give up on reading the channel and not block the
+	// child goroutine.
+	errC := make(chan error, 1)
+	go func() {
```

Comment:
> Why do we need this goroutine?

Isn't it because it allows the calling goroutine to give up early and not wait? For example, does your proposal handles the situation where `fn()` is blocked due to filesystem I/O, maybe over the network with potential high latency and timeouts?

Reply:
ah I see what you mean